clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -1871,4 +1873,100 @@ void handleFailure(Throwable throwable) {
         return new CreatePartitionsResult(new HashMap<String, KafkaFuture<Void>>(futures));
     }
 
+    public DeleteRecordsResult deleteRecords(final Map<TopicPartition, RecordsToDelete> partitionsAndOffsets,
+                                             final DeleteRecordsOptions options) {
+
+        // requests need to be sent to partitions leader nodes so ...
+        // ... from the provided map it's needed to create more maps grouping topic/partition per leader
+
+        final Map<TopicPartition, KafkaFutureImpl<Long>> futures = new HashMap<>(partitionsAndOffsets.size());
+        for (TopicPartition topicPartition: partitionsAndOffsets.keySet()) {
+            futures.put(topicPartition, new KafkaFutureImpl<Long>());
+        }
+
+        // preparing topics list for asking metadata about them
+        final Set<String> topics = new HashSet<>();
+        for (TopicPartition topicPartition: partitionsAndOffsets.keySet()) {
+            topics.add(topicPartition.topic());
+        }
+
+        final long nowMetadata = time.milliseconds();
+        final long deadline = calcDeadlineMs(nowMetadata, options.timeoutMs());
```

Comment:
Hmm.. we already have a `metadata` object that is keeping updated by the `AdminClientRunnable`, can we just call `metadata.fetch()` to get the current cluster information? Then in line 1918 if we do not have the current leader we can still return `LEADER_NOT_AVAILABLE` to let the caller retry as it is a retryable error code.

Reply:
I don't think that will work, because the metadata object that is maintained in AdminClientRunnable does not fetch information about any topics.  In general the way the Metadata object works is a very poor fit for AdminClient-- it was designed for producers and consumers, where you have long-running subscriptions and so forth.  It is much simpler and better just to make the metadata call here-- we can always optimize this later with a cache if needed.

Reply:
@guozhangwang it was the way I started with the implementation but then I noticed what @cmccabe is saying (no topics information fetched). My first way was to do a metadata request with topics in a similar way the producer does before using the `AdminClientRunnable` for the delete records call. I didn't like it too much (even because it was synchronous) so I switched to use the `AdminClientRunnable` for both (metadata and delete requests) leveraging on what is already working well today.